### PR TITLE
d/virtual_machine: Disk information is now a sub-resource

### PR DESCRIPTION
**Pull request by vancluever**
This is a follow up to #277 which adds extra validation to the cloning process.

This updates the `vsphere_virtual_machine` so that instead of `disk_sizes`, there is now a `disks` sub-resource that not only contains the size of each disk, but also the settings for `eagerly_scrub` and `thin_provisioned`. This ensures that these attributes can be sourced from the template and set correctly when deploying linked clones.

Tests have been updated, in addition to the docs. There are also new tests that validate that errors are triggered on invalid clone options.

Will be posting the results of the entire VM test suite when it completes.

### CODE REVIEW

**Comment by vancluever** [comment]
Tests done:

```
=== RUN   TestAccResourceVSphereVirtualMachine
=== RUN   TestAccResourceVSphereVirtualMachine/basic
=== RUN   TestAccResourceVSphereVirtualMachine/ESXi-only_test
=== RUN   TestAccResourceVSphereVirtualMachine/shutdown_OK
=== RUN   TestAccResourceVSphereVirtualMachine/multi-device
=== RUN   TestAccResourceVSphereVirtualMachine/add_devices
=== RUN   TestAccResourceVSphereVirtualMachine/remove_middle_devices
=== RUN   TestAccResourceVSphereVirtualMachine/remove_middle_devices,_change_disk_unit
=== RUN   TestAccResourceVSphereVirtualMachine/high_disk_unit_numbers
=== RUN   TestAccResourceVSphereVirtualMachine/high_disk_units_to_regular_single_controller
=== RUN   TestAccResourceVSphereVirtualMachine/cdrom
=== RUN   TestAccResourceVSphereVirtualMachine/maximum_number_of_nics
=== RUN   TestAccResourceVSphereVirtualMachine/upgrade_cpu_and_ram
=== RUN   TestAccResourceVSphereVirtualMachine/modify_annotation
=== RUN   TestAccResourceVSphereVirtualMachine/grow_disk
=== RUN   TestAccResourceVSphereVirtualMachine/swap_scsi_bus
=== RUN   TestAccResourceVSphereVirtualMachine/extraconfig
=== RUN   TestAccResourceVSphereVirtualMachine/extraconfig_swap_keys
=== RUN   TestAccResourceVSphereVirtualMachine/attach_existing_vmdk
=== RUN   TestAccResourceVSphereVirtualMachine/in_folder
=== RUN   TestAccResourceVSphereVirtualMachine/move_to_folder
=== RUN   TestAccResourceVSphereVirtualMachine/static_mac
=== RUN   TestAccResourceVSphereVirtualMachine/single_tag
=== RUN   TestAccResourceVSphereVirtualMachine/multiple_tags
=== RUN   TestAccResourceVSphereVirtualMachine/switch_tags
=== RUN   TestAccResourceVSphereVirtualMachine/clone_from_template
=== RUN   TestAccResourceVSphereVirtualMachine/clone,_multi-nic_(template_should_have_one)
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_different_timezone
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_bad_timezone
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_bad_eagerly_scrub_with_linked_clone
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_bad_thin_provisioned_with_linked_clone
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_bad_size_with_linked_clone
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_bad_size_without_linked_clone
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_different_hostname
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_extra_disks
=== RUN   TestAccResourceVSphereVirtualMachine/clone_with_cdrom
=== RUN   TestAccResourceVSphereVirtualMachine/cpu_hot_add
=== RUN   TestAccResourceVSphereVirtualMachine/memory_hot_add
=== RUN   TestAccResourceVSphereVirtualMachine/dual-stack_ipv4_and_ipv6
=== RUN   TestAccResourceVSphereVirtualMachine/ipv6_only
=== RUN   TestAccResourceVSphereVirtualMachine/windows_template,_customization_events_and_proper_IP
=== RUN   TestAccResourceVSphereVirtualMachine/host_vmotion
=== RUN   TestAccResourceVSphereVirtualMachine/resource_pool_vmotion
=== RUN   TestAccResourceVSphereVirtualMachine/storage_vmotion_-_global_setting
=== RUN   TestAccResourceVSphereVirtualMachine/storage_vmotion_-_single_disk
=== RUN   TestAccResourceVSphereVirtualMachine/storage_vmotion_-_pin_datastore
=== RUN   TestAccResourceVSphereVirtualMachine/import
=== RUN   TestAccResourceVSphereVirtualMachine/import_with_multiple_disks_at_different_SCSI_slots
--- PASS: TestAccResourceVSphereVirtualMachine (8055.66s)
    --- PASS: TestAccResourceVSphereVirtualMachine/basic (163.74s)
    --- SKIP: TestAccResourceVSphereVirtualMachine/ESXi-only_test (0.00s)
    	helper_test.go:85: set VSPHERE_TEST_ESXI to run ESXi-specific acceptance tests
    --- PASS: TestAccResourceVSphereVirtualMachine/shutdown_OK (155.27s)
    --- PASS: TestAccResourceVSphereVirtualMachine/multi-device (173.95s)
    --- PASS: TestAccResourceVSphereVirtualMachine/add_devices (178.01s)
    --- PASS: TestAccResourceVSphereVirtualMachine/remove_middle_devices (180.20s)
    --- PASS: TestAccResourceVSphereVirtualMachine/remove_middle_devices,_change_disk_unit (303.07s)
    --- PASS: TestAccResourceVSphereVirtualMachine/high_disk_unit_numbers (170.23s)
    --- PASS: TestAccResourceVSphereVirtualMachine/high_disk_units_to_regular_single_controller (320.57s)
    --- PASS: TestAccResourceVSphereVirtualMachine/cdrom (19.69s)
    --- PASS: TestAccResourceVSphereVirtualMachine/maximum_number_of_nics (180.46s)
    --- PASS: TestAccResourceVSphereVirtualMachine/upgrade_cpu_and_ram (290.87s)
    --- PASS: TestAccResourceVSphereVirtualMachine/modify_annotation (177.79s)
    --- PASS: TestAccResourceVSphereVirtualMachine/grow_disk (172.81s)
    --- PASS: TestAccResourceVSphereVirtualMachine/swap_scsi_bus (290.64s)
    --- PASS: TestAccResourceVSphereVirtualMachine/extraconfig (159.61s)
    --- PASS: TestAccResourceVSphereVirtualMachine/extraconfig_swap_keys (290.63s)
    --- PASS: TestAccResourceVSphereVirtualMachine/attach_existing_vmdk (168.99s)
    --- PASS: TestAccResourceVSphereVirtualMachine/in_folder (161.77s)
    --- PASS: TestAccResourceVSphereVirtualMachine/move_to_folder (170.66s)
    --- PASS: TestAccResourceVSphereVirtualMachine/static_mac (169.31s)
    --- PASS: TestAccResourceVSphereVirtualMachine/single_tag (171.35s)
    --- PASS: TestAccResourceVSphereVirtualMachine/multiple_tags (167.59s)
    --- PASS: TestAccResourceVSphereVirtualMachine/switch_tags (172.76s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_from_template (120.00s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone,_multi-nic_(template_should_have_one) (127.70s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_different_timezone (120.22s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_bad_timezone (0.01s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_bad_eagerly_scrub_with_linked_clone (5.80s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_bad_thin_provisioned_with_linked_clone (5.09s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_bad_size_with_linked_clone (5.41s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_bad_size_without_linked_clone (5.02s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_different_hostname (129.96s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_extra_disks (285.28s)
    --- PASS: TestAccResourceVSphereVirtualMachine/clone_with_cdrom (126.42s)
    --- PASS: TestAccResourceVSphereVirtualMachine/cpu_hot_add (140.02s)
    --- PASS: TestAccResourceVSphereVirtualMachine/memory_hot_add (140.78s)
    --- PASS: TestAccResourceVSphereVirtualMachine/dual-stack_ipv4_and_ipv6 (120.42s)
    --- PASS: TestAccResourceVSphereVirtualMachine/ipv6_only (429.83s)
    --- PASS: TestAccResourceVSphereVirtualMachine/windows_template,_customization_events_and_proper_IP (242.14s)
    --- PASS: TestAccResourceVSphereVirtualMachine/host_vmotion (146.79s)
    --- PASS: TestAccResourceVSphereVirtualMachine/resource_pool_vmotion (132.53s)
    --- PASS: TestAccResourceVSphereVirtualMachine/storage_vmotion_-_global_setting (390.75s)
    --- PASS: TestAccResourceVSphereVirtualMachine/storage_vmotion_-_single_disk (258.15s)
    --- PASS: TestAccResourceVSphereVirtualMachine/storage_vmotion_-_pin_datastore (392.22s)
    --- PASS: TestAccResourceVSphereVirtualMachine/import (158.95s)
    --- PASS: TestAccResourceVSphereVirtualMachine/import_with_multiple_disks_at_different_SCSI_slots (162.18s)
```

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 1]
```diff
@@ -4050,17 +4116,15 @@ resource "vsphere_virtual_machine" "vm" {
       linux_options {
         host_name = "terraform-test"
         domain    = "test.internal"
+				time_zone = "${var.time_zone}"
```
Comment (tombuildsstuff) [inline vsphere/resource_vsphere_virtual_machine_test.go]: _minor_ indentation

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 2]
```diff
@@ -4258,21 +4349,27 @@ data "vsphere_virtual_machine" "template" {
...
   disk {
-    name = "terraform-test.vmdk"
-    size = "${data.vsphere_virtual_machine.template.disk_sizes[0]}"
+    name             = "terraform-test.vmdk"
+    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
+    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
+    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
+  }
+  
+	cdrom {
```
Comment (tombuildsstuff) [inline vsphere/resource_vsphere_virtual_machine_test.go]: _minor_ indentation

File: vsphere/resource_vsphere_virtual_machine_test.go [hunk 3]
```diff
@@ -4504,16 +4613,500 @@ resource "vsphere_virtual_machine" "vm" {
...
+        host_name = "terraform-test"
+        domain    = "test.internal"
+      }
+
+      network_interface {
+				ipv6_address = "fd00::2"
+        ipv6_netmask = "32"
+      }
+
+			ipv6_gateway = "fd00::1"
```
Comment (tombuildsstuff) [inline vsphere/resource_vsphere_virtual_machine_test.go]: _minor_ indentation

**Comment by vancluever** [comment]
Thanks @tombuildsstuff!

This is the data source, so there's no state to migrate, people will just need to change their configuration.  This will be noted as a breaking change in the changelog, but hopefully this is quick enough since release that we can get this stable interface in for the forseeable future.

Will fix the indents.

**Pull request merged**